ipa-core/src/cli/playbook/mod.rs

```diff
@@ -46,7 +47,75 @@ where
             break;
         }
 
-        let same = next_expected == next_actual;
+        let same = next_expected == next_actual; // with DP non-exact match here
+
+        let color = if same { Color::Green } else { Color::Red };
+        table.add_row(vec![
+            Cell::new(format!("{i}")).fg(color),
+            Cell::new(format!("{next_expected:?}")).fg(color),
+            Cell::new(format!("{next_actual:?}")).fg(color),
+            Cell::new(if same { "" } else { "X" }),
+        ]);
+
+        if !same {
+            mismatch.push((i, next_expected, next_actual));
+        }
+
+        i += 1;
+    }
+
+    tracing::info!("\n{table}\n");
+
+    assert!(
+        mismatch.is_empty(),
+        "Expected and actual results don't match: {mismatch:?}",
+    );
+}
+
+/// Validates that the expected result matches the actual.
+///
+/// ## Panics
+/// If results don't match.
+pub fn validate_dp(expected: Vec<u32>, actual: Vec<u32>, epsilon: f64, per_user_credit_cap: u32) {
```

Comment:
this seems to replicate the behavior of `validate` almost entirely except the part where you compare the values. Can you make `validate` generic to work for both cases? It seems that you should be able to by computing the `std` and passing it as optional delta to the generic function

Reply:
Look at it a little and I don't think it could work with the current trait bounds on `validate` as `PartialEq` only supports equal and not equal comparisons. But with DP we would need some <=, >= operations.  

Likely the simplest way to duplicate the code is to use the `validate_dp` function for both.   

Either direction, I'd like to push off to some future PR.